man/ldpl.1
.TH LDPL 1 "5 may 2019" "LDPL Man 1.0"

.SH NAME
ldpl - The LDPL programming language compiler

.SH SYNOPSIS
.PP
ldpl [-i=<extension file>]... <source file>|-c
     [-o=<output name>|-r] [-f=<c++ flag>]... [-n]
.br
ldpl [-v|-h]

For more information on these options you can run "ldpl -h".

.SH DESCRIPTION
.PP
This program compiles LDPL source code into executable binaries.
LDPL is a programming language designed from the ground up to be excessively expressive, fast, readable and easy to learn.
In order to be able to properly compile LDPL executables you
.I
must
have a valid C++ compiler on your $PATH linked to the name 'c++'.

Documentation and reference for the LDPL Programming Language can be found in the sections below.

.SH OPTIONS
.PP
-v, --version
    Print out LDPL version info and release details.

-h, --help
    Print this list of options.

-r
    By using -r you can just compile the project and print the C++ representation for that code.

-o=<name>
    You can set the output file for the compiled binary with the -o flag. For example, if you want to name your program "dog", you could compile it with ldpl -o=dog main.ldpl.

-i=<file>
    Extensions can be imported by using this flag; see the Extensions section.

-f=<flag>
    The -f flag can be used to add flags to the C++ compilation line. See the Building C++ Extensions section for more information.

-n, --non-static
    On Linux and Windows platforms, LDPL builds static binaries by default. If you want to build non-static ones use the -ns flag. To build on Android Termux you must use this flag.

-c
    The -c flag tells LDPL to accept source code from the standard input.

.SH AUTHOR
This document is based on the LDPL Reference, written by Martín del Río in collaboration with Chris West.

.SH REPORTING BUGS
Report any bugs to <https://github.com/lartu/ldpl>.

.SH COPYRIGHT
Copyright © 2018 - 2019, Martín del Río. LDPL may be copied only under the terms of the GNU General Public License 3.0, which may be found in the LDPL repository.
.br
This is free software: you are free to change and redistribute it. There is NO WARRANTY, to the extent permitted by law.

.ce 1
.SH -=-=-=-=-=-=-=-=-=- DOCS: LDPL DOCUMENTATION -=-=-=-=-=-=-=-=-=-


# LDPL Docs

## Introduction

The LDPL Community has compiled this document with the desire to teach
and standardize the LDPL programming language. This document contains the
specification for the LDPL programming language, as well as explanations,
instructions and examples for every feature included in it.


Feedback, corrections and suggestions are welcome, both on the main
LDPL repository
or by e-mail to martin@ldpl-lang.org.
You can join the LDPL community at r/ldpl
or via IRC on irc.freenode.net channel #ldpl.

The source code for this documentation is available in the LDPL repository.

## About LDPL

LDPL is a powerful compiled programming language designed from the
ground up to be expressive, readable, fast and easy to learn. It mimics plain
English, in the likeness of the good parts of older programming languages like
COBOL, with the desire that it can be understood by anybody. LDPL was designed
to run on Unix systems, including AMD-64 Linux, macOS, ARMv8 Linux, Android
Termux and both Intel and PowerPC OS X (tested from Tiger 10.4 onwards).
It even supports UTF-8 out of the box.

    # LDPL 'Hello World' example

    data:
    name is text # Your name will go here.

    procedure:
    display "Hello World!" lf "What's your name? "
    accept name
    display "你好, " name ", welcome to LDPL!" lf

LDPL also aims to suppress unreadable code and redundancy by only having one
way of doing anything. What a command does should never overlap with the
purpose of another command and, as such, every LDPL command does one and only
one thing. Every line is a step further towards the completion of an algorithm,
no more, no less.

## The LDPL Compiler

To use LDPL, you should first download or compile the LDPL compiler.
For more information on how to do that, read the
*How to install LDPL* section
on the LDPL Readme.

To use the compiler, you must have a C++ compiler already installed on your
system and have mapped it to 'c++', found on your PATH. The LDPL Compiler
compiles LDPL code to C++ code and thus this is a requirement for it to work.

Once the compiler is set up, go write some LDPL source code, say 'source.ldpl'.
Then compile the source code using 'ldpl source.ldpl'. The compiled, executable
binary file will be saved as 'source-bin'. Done!

### Compiler Switches

To use the LDPL compiler, the syntax is as follows:

    ldpl [-i='<included file>']... <source file>|-c
         [-o='<output name>'|-r] [-f='<c++ flag>']... [-n]
    ldpl [-v|-h]

The '-f' flag can be used to pass extra options to the 'c++' compiler when. For
example, '-f=-lSDL' could be used to link against SDL. The 'flag' statement can
also be used as well ('flag "-lSDL"') and its use is recommended. More on this
later.

By using '-r' you can just compile your project and print the C++ representation
for that code.

You can set the filename for the compiled binary with the '-o' flag. For
example, if you want to name your program "dog", you could compile it with
'ldpl -o=dog main.ldpl'.

On Linux platforms, LDPL builds static binaries by default. If you want to
build non-static ones use the '-ns' flag.

The '-c' flag tells LDPL to accept source code from the standard input.


You can import extra files and extensions to your LDPL compilation by using the
'-i=' flag. Extensions can be imported by passing '.o', '.a', or '.cpp' files
to this flag; see the Extensions section for more information. The use of the
'INCLUDE' statement is preferred.

'-v' and '--version' print out version info and release details.

'-h' and '--help' print this list of options.

## File Extensions

The preferred file extension for LDPL source files is '.ldpl'.
The extension '.lsc' (LDPL Source File) should also be accepted in case
the preferred extension couldn't be used for any reason.

~ Tip ~
    File extensions are important: they help editors to recognize what language
    your source code is written in and they tell the LDPL compiler how to treat
    your files.

## License

This LDPL Compiler is distributed under the Apache 2.0 License.
All LDPL Dinosaur logos were created by Lartu and are
released under a Creative Commons Attribution 4.0 International (CC BY 4.0)
license. This documentation is released under the Apache 2.0 License.




# LDPL Source Code Structure

## Case Sensitivity

LDPL is a case-insensitive language. This means that variables and statements
that consist of the same characters are understood to be the same variable
or statement regardless of their case. For example, in other languages the
variables 'foo' and 'FOO' would represent two different variables, but in
LDPL they are the same one. This same thing happens with statements. For
LDPL it's the same if you write 'display' or 'dIsPlAy' (but please don't do so).

~ Warning ~
    LDPL is case-insensitive only for latin A-z characters. Accented characters and
    non-ASCII letters will be matched in a case-insensitive manner. Depending
    on the LDPL implementation, 'ááá' and 'ÁÁÁ' may not represent the same
    identifier. It is good practice to write your code in a single case and
    always call your variables and functions by the exact name used when
    declaring them.

## Comments
Comments in LDPL are denoted with a hash symbol ('#') and can be placed both
on their own line or at the end of a line that already contains a statement.
Everything after the hash is considered to be part of the comment and,
therefore, not executed.

    data:
        # This won't be executed
        # This won't be either
    procedure:
        # This line won't be interpreted

## Data and Procedure Sections
LDPL was designed to be a rigidly structured programming
language and, as such, variable declarations and the remaining code procedures
are separated into two different, mutually exclusive sections within every
source file: the 'data:' section and the 'procedure:' section.

    data:
        # Variables go here
    procedure:
        # The rest of the code goes here

Variable declarations should be placed within the data section, while the
rest of the code should be placed inside the procedure section. Further
sub-procedures should be placed also within the procedure section, inside their
own sub-procedure subsections.

The data section may be obviated if no variables are used.

If your project consists of multiple LDPL source files, each file must have its
own data and procedure sections.

## Including More Source Files

You can import other LDPL source files to your LDPL source by using the
include statement. For example, say you have two sources:

    # This is 'firstSource.ldpl'
    procedure:
        call someSubprocedure

and

    # This is 'includedFile.ldpl'
    procedure:
        sub someSubprocedure
            display "Hi there!"
        end sub

You can import the second source into the first one in order to create one
big source file like this:

    # This is 'firstSource.ldpl'
    include "includedFile.ldpl"
    procedure:
        call someSubprocedure

When you run the code above, it will print

    Hi there!

using the sub-procedure 'someSubprocedure' included from the second file.

The location where the included files are searched for is relative to the file
that includes them. You may include as many files as you like.

The 'include' statement can only be used before the data section.

## C++ Extensions

LDPL source code can be extended using C++ Extensions, C++ source files
that can be compiled along the C++ source code generated by the LDPL compiler.
While these are explained in greater detail in their respective section of
this document, two statements are relevant to this part of the documentation:
the 'flag' and 'extension' statements.

## C++ Compiler Flags

When writing C++ code, you may need to pass some flags to the C++ compiler.
Say, for example, you are writing something using the SDL Library. When
trying to compile your code, you will need to pass the flag '-lSDL' to your
C++ compiler. This same thing can be achieved under LDPL by using the 'flag'
statement.

Following the example above, if you want to pass the flag '-lSDL' to the C++
compiler that compiles the code generated by the LDPL compiler from your LDPL
source code, you may do this:

    flag "-lSDL"
    flag "-fpermissive" # you can pass as many flags as you want
    data:
        #...
    procedure:
        #...

Often, different operating systems require different flags. In those cases,
you can restrict the 'flag' statement to a particular OS:

    flag "-O3"
    flag linux "-lncursesw"
    flag macos "-lncurses"
    flag macos "-D_XOPEN_SOURCE_EXTENDED" # multiple flags per OS are permitted
    data:
        #...
    procedure:
        #...

~ Tip ~
    LDPL officially supports the following operating systems:

    • Linux
    • MacOS
    • Android

    And includes experimental support for these fine systems:

    • BSD
    • Emscripten

The 'flag' statement can only be used before the data section.

## C++ Extension Including

Extensions are C++ source files that you can compile along your LDPL source
in order to extend the language, as stated above. If you want to include an
extension, you may use the 'extension' statement. For example:

    extension "someDirectory/someFile.cpp"
    flag "-O3"
    data:
        #...
    procedure:
        #...

The 'extension' statement can only be used before the data section.









## The Data Section

The data section is where global variables are declared (you can use them
anywhere in your program). If no variables are declared, the data section can
be skipped altogether.

The data section is defined and preceded by the 'data:' keyword.
An empty data section looks like this:

    data:

On every line within the data section (that is, on every line after the 'data:'
keyword and before the 'procedure' keyword) one and only one variable can be
declared.

The syntax for declaring a variable in LDPL is:

    variable-name is data-type

Variable naming schemes and data types will be explained later on this
document.

A data section cannot contain anything but variable declarations, comments and
empty lines. An example data section may end up looking like this:

    data: # this is an example data section
        foo is number
        bar is text map

        # foobar is my number list
        foobar is number list

## Variable Data Types

LDPL natively supports the scalar number and text data types. It also
supports containers of said scalar types: maps and lists, combined in
any way.

Variables may be declared, as stated above, using the syntax:

    variable-name is data-type

within the data section. The name of a data type is composed of either a scalar
data type name (number or text) or a container type plus one or more
scalar or container type names. For example:

    foo is number
    bar is text
    foobar is map of lists of text

In the case above, foobar is a map of lists of text values.

### The Number Data Type

The number data type, as its name suggests, depicts numeric values.
It's recommended that it be represented internally as a binary64
double-precision floating-point format number as defined by the IEEE 754.

Both variables and numeric constants can be members of the number type.

Valid number literals must begin with a decimal value (for example 5 or 0.12,
.12 wouldn't be a valid number) and may be preceded by a minus sign for
negative numbers (-5, -567.912). Numbers may not be preceded by a plus sign
(+5 is not a valid number literal). The literal -0 is implicitly transformed
into 0.

    -231897.123 # is an example number

### The Text Data Type

The text data type, as its name suggests, represents alphanumeric strings.
In the interest of supporting as many locales as possible, texts should be utf-8
encoded to be compatible with Unicode. A text maximum length for text values is
explicitly not defined and it should be limited only by the amount of available
memory on the system. Strings in LDPL are enclosed between two "quotes".

    "This is an example string!"

LDPL strings may contain multiple escape sequences / control characters in
them. Each escape sequence counts as only one character. The available escape
sequences are:

    • '\\a' = alert (bell)
    • '\\b' = backspace
    • '\\t' = horizontal tab
    • '\\n' = newline / line feed
    • '\\v' = vertical tab
    • '\\f' = form feed
    • '\\r' = carriage return
    • '\\e' = non-standard GCC escape
    • '\\0' = null byte
    • '\\\\' = \\ character
    • '\\"' = " character

For example, the string '"hello,\\nworld"' will be displayed as

    hello,
    world

when printed to the console.

### The List Data Type

The list data type is a collection of number or text values, or containers
of possibly more containers of said types. Values can be pushed to lists and
then be accessed and modified using the ':' operator. List indexes consist of
integer numbers. The first index of a list is index 0, and the rest count up to
the length of the list minus one.

Lists, as collections of number or text values (or collections of said types),
can only have one defined type at any given time: text or number. A single list
is not capable of storing both numeric and alphanumeric values.

    data:
        foo is number list # This is a list

    procedure:
        push 10 to foo
        display foo:0 lf

~ Note ~
    The 'display' statement prints values to the screen. While it will be
    explained in more detail later, the line

        'display foo:0 lf'

    prints
    the value of 'foo:0' (the index '0' of the list 'foo') followed by
    a line break ('lf').

In most other languages, you may have to declare a list and specify how many
elements or components it contains. In such languages, the declaration causes a
contiguous block of memory to be allocated for that many elements. LDPL lists,
however, are different: they are dynamic. This means that you can store as many
values as you want in a single list without fear of running out of place (of
course, this is limited by the memory allocated by your operating system for
your LDPL program).

Suppose you store the values '"hi"', '"there"', '"I love"' and
'"LDPL, it's great!"' in a 'text list', in that particular order. Then the
contents of the list and the index associated with each element will be

.br
| Index | Element |
.br
| :---: | :---: |
.br
| 0 | '"hi"' |
.br
| 1 | '"there"' |
.br
| 2 | '"I love"' |
.br
| 3 | '"LDPL, it's great!"' |

We have shown the pairs in order because their order is relevant: if you added
a new element to the list, it would be inserted after the last element, thus
being associated with index '4'.

To add values to a list, you must first push them to the list. For example,
if you want to add the numbers '10', '20' and '30' to a 'number list', your
code should look like this:

    data:
      myList is number list
    procedure:
      push 10 to myList # 10 is stored in index 0 of myList
      push 20 to myList # 20 is stored in index 1 of myList
      push 30 to myList # 30 is stored in index 2 of myList


Values in lists can be stored and accessed just like any other variable
(see the store - in statement for further details) using the ':' operator.
This operator indicates what index of the list we are writing to or reading
from. Here we declare a 'number list' and store the values '5' and '-10.2' in
it, and then replace the number '5' by the number '890':

    data:
      myList is number list
    procedure:
      push 5 to myList
      # 5 is stored in index 0 of myList
      push -10.2 to myList
      # -10.2 is stored in index 1 of myList
      store 890 in myList:0
      # We store 890 in index 0 of myList, thus replacing the 5

~ Note ~
    The 'push' statement adds a value at the end of a list. Also, the 'store
    • in' statement stores a value in a variable. These two statements will
    be explained in more detail in their respective sections of this document.

Please note that as a list is variable that's a collection of values, a
single index of a list is a variable in itself. This means that any subindex
of a list that resolves to a scalar value be used in any position where you
could use a variable of the same type of that value. So, if you have something
like this:

    store <number-variable or number> in <number-variable>
.br

You could use a 'number list' with a defined sub-index \(for example, in the
example above, 'myList:0'\) where it says number-var, just like in
the 'store - in' examples in the code extracts above.

In the list statements section, you'll find a collection of statements that
can be used to work with lists.

### The Map Data Type

The map data type is a collection of number or text values. Maps
superficially resemble lists but with fundamental differences. The biggest one
is that any number or string in LDPL may be used as an array index, not just
consecutive integers. Also, values in maps have no order.

    data:
        foo is number map # This is a map

    procedure:
        store 19 in foo:"hi there!"
        display foo:"hi there!" lf

Maps, as collections of number or text values (or collections of said types),
can only have one defined type at any given time: text or number. A single map
is not capable of storing both numeric and alphanumeric values.

Unlike lists, maps are associative. This means that each map is a
collection of pairs: a key and its corresponding element. For example,
you could have a 'number map' with the following contents:

.br
| Key | Element |
.br
| :---: | :---: |
.br
| 4 | 30 |
.br
| 2 | 10 |
.br
| "Hi there!" | -56.3 |
.br
| "99ldplrocks89" | 0 |

We have shown the pairs in jumbled order because their order is irrelevant.
One advantage of maps is that new pairs can be added at any time. maps can be
sparse: they can have missing keys \(say for example you have keys 1 and 5, but
don't have keys 2, 3 and 4\). Another consequence of maps is that the keys
don't necessarily have to be positive integers. Any number, or even a string,
can be a key.

Values in maps can be stored and accessed just like any other variable
(see the store - in statement for further details) using the ':' operator.
This operator indicates what key of the map we are writing to or reading from.
Here we declare a 'number map' and store the values '5' and '-10.2' in the
keys '1' and '5', respectively.

    data:
      myMap is number map
    procedure:
      store 5 in myMap:1 #Stores 5 in key 1 of myMap
      store -10.2 in myMap:5 #Stores -10.2 in key 5 of myMap

As stated before, map keys don't always have to be constant numbers.
They can also be number variables, text and text variables, or even sub-indexes
of lists or elements from other maps. For example:

    data:
      myMap is number map
      myOtherMap is number map
      myVar is number

    procedure:
      store 17 in myVar
      store 1 in myMap:"hello"
      #Stores 1 in key "hello" of myMap
      store 7 in myMap:myVar
      #Stores 7 in a key equal to the current value of myVar
      store 3 in myMap:myOtherMap:4
      #Stores 3 in a key of equal value to the key of myMap with value equal to
      #key 4 of myOtherMap

In fact, when you use a number value as a subindex for a map, it is silently
casted into a text value. For example, 'myMap:1' will be interpreted \(an thus,
the same\) as 'myMap:"1"'.

Please note that as a map is variable that's a collection of values,
a single key of a map is a variable in itself. This means that any key of a
map that resolves to a scalar value can be used in any position where you could
use a variable of the same type of that value. So, if you have something like
this:

    store <number-var or number> in <number-var>
.br

You could use a 'number map' with a particular key where it says number-var,
just like in the 'store - in' examples in the code extracts above
\(for example 'myMap:"hello"'\).

As you'll see in the Default Variable Values section, you can access
undeclared keys of a map, just like if they were declared.
See the following example:

    data:
.br
      myMap is number map
.br
    procedure:
.br
      display myMap:99
.br

In the example above, '0' will be printed and no errors displayed during
compilation, even though the key '99' of 'myMap' hasn't been explicitly
declared. This is because when you try to access an element that hasn't
been declared yet, LDPL declares it for you and initializes it to its type
default value.

It's important to note that this very feature is a double-edged weapon. While
you can use it to access uninitialized map keys, you cannot check if a value
exists in a map without initializing it if it wasn't there before. Statements
like store key count of and store keys of are provided as means to
overcome this situation.


src/ldpl.cpp
// +=============================================+
// |                                             |
// |          The __    ____  ____  __           |
// |             / /   / __ \/ __ \/ /           |
// |            / /   / / / / /_/ / /            |
// |           / /___/ /_/ / ____/ /___          |
// |          /_____/_____/_/   /_____/          |
// |               Programming Language          |
// |          http://www.ldpl-lang.org/          |
// |                                             |
// +=============================================+

// +------------------+
// | --- Includes --- |
// +------------------+
#include "ldpl.h"
#include "aux/aux_c_format.cpp" // Contains auxiliary functions that turn LDPL identifiers / expressions into C / C++ identifiers / expressions
#include "aux/aux_code.cpp" // Contains auxiliary functions that add C++ code to the generated C++ file
#include "aux/aux_compile_line.cpp" // Contains the behemoth compile_line function that pattern-matches lines and compiles them
#include "aux/aux_container.cpp" // Contains auxiliary functions to split and access [multi]containers
#include "aux/aux_format.cpp" // Contains auxiliary functions that do text or number formatting
#include "aux/aux_info.cpp" // Contains auxiliary functions that display information about the compiler and the system
#include "aux/aux_line_like.cpp" // Contains the line_like function used for pattern matching lines
#include "aux/aux_state.cpp" // Contains auxiliary functions that check the current compilation state
#include "aux/aux_tokenizer.cpp" // Contains functions used to tokenize an LDPL source file
#include "aux/aux_typecheck.cpp" // Contains auxiliary functions to check data type compliance
#include <memory>

// +----------------------------+
// | --- Main Aux Functions --- |
// +----------------------------+

string exec(const char *cmd)
{
    // -- Executes a command passed as a string -- 
    array<char, 128> buffer;
    string result;
    unique_ptr<FILE, decltype(&pclose)> pipe(popen(cmd, "r"), pclose);
    if (!pipe)
    {
        throw runtime_error("popen() failed!");
    }
    while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr)
    {
        result += buffer.data();
    }
    return result;
}

void load_and_compile(string &filename, compiler_state &state)
{
    // -- Loads a file and sends it to compilation --
    filename = expandHomeDirectory(filename);
    state.where.current_file = filename;
    // Accept input from stdin
    ifstream file(filename);
    // Fail if the file couldn't be loaded
    if (!file.is_open())
        error("The file '" + filename + "' couldn't be loaded.");
    // Get file contents
    vector<string> lines;
    string line = "";
    while (getline(file, line))
    {
        replace_whitespace(line);
        lines.push_back(line);
    }
    bullet_msg("Compiling " + filename);
    compile(lines, state);
}

void accept_and_compile(compiler_state &state)
{
    // -- Accepts LDPL source code via standard input and sends it to compilation --
    // Get file contents
    vector<string> lines;
    state.where.current_file = "standard input";
    string line = "";
    while (getline(cin, line))
    {
        replace_whitespace(line);
        lines.push_back(line);
    }
    bullet_msg("Compiling");
    compile(lines, state);
}

void compile(vector<string> &lines, compiler_state &state)
{
    // -- Takes a vector of LDPL source lines and compiles them to C++ --
    // For each line in the source code
    for (size_t line_num = 1; line_num <= lines.size(); ++line_num)
    {
        string &line = lines[line_num - 1];
        state.where.current_line = line_num;

        if (state.open_quote)
        {
            // Check for END QUOTE first
            if (line.size() >= 9 /*&& (line[0] == 'E' || line[0] == 'e')*/)
            {
                string upper = "";
                for (char c : line)
                    upper += toupper(c);
                trim(upper);
                if (upper == "END QUOTE")
                {
                    state.open_quote = false;
                    // Kill final newline. Programs can add crlf if needed.
                    string &prev = state.current_subprocedure != "" ? state.subroutine_code.back() : state.output_code.back();
                    size_t pos = prev.rfind("\\n");
                    if (pos != string::npos)
                        prev.erase(pos, 2);
                    prev += ";";
                    continue;
                }
            }

            // No END QUOTE, emit the line as C++
            state.add_code("\"" + escape_c_quotes(line) + "\\n\"", state.where);
            continue;
        }

        trim(line);
        // Split line in various tokens
        vector<string> tokens;
        tokenize(line, tokens, state.where, true, ' ');
        for (string &token : tokens)
            if (token == "CRLF" || token == "LF")
                token = "\"\\n\"";
        if (tokens.size() == 0)
            continue;
        compile_line(tokens, state);
    }
    if (state.open_quote)
        error("a QUOTE block was not terminated.");
    if (state.closing_subprocedure())
        error("a SUB-PROCEDURE block was not terminated.");
    if (state.closing_if())
        error("a IF block was not terminated.");
    if (state.closing_loop())
        error("a WHILE or FOR block was not terminated.");
}

// +-----------------------+
// | --- Main Function --- |
// +-----------------------+

int main(int argc, const char *argv[])
{
    // Get command line arguments as string vector
    vector<string> args(argv + 1, argv + argc);

    // Check if the user requested version or help information
    if (contains_any(args, {"-v", "--version"}))
    {
        displayVersionInfo();
        return 0;
    }
    else if (contains_any(args, {"-h", "--help"}))
    {
        displayHelpInfo();
        return 0;
    }

    // Define local variables that will be used for compilation
    compiler_state state; // Compiler state (holds variables, sections, functions, etc)
    vector<string> files_to_compile;
#ifdef STATIC_BUILDS
    bool no_static = false;
#endif
    string output_filename = "";
    string final_filename = "";

    // Check command line arguments
    if (args.size() >= 1)
    {
        for (string &arg : args)
        {
            if (arg.size() >= 1 && arg[0] != '-')
            {
                if (output_filename == "")
                    output_filename = arg;
                if (files_to_compile.size() > 0)
                {
                    warning("passing multiple LDPL source files to the\ncompiler is deprecated and may be removed in the future.\nPlease use the IMPORT statement instead.");
                }
                files_to_compile.push_back(arg);
            }
            else if (arg == "-r")
            {
                show_ir = true;
            }
#ifdef STATIC_BUILDS
            else if (arg == "-n" || arg == "--non-static")
            {
                no_static = true;
            }
#endif
            else if (arg.substr(0, 3) == "-o=")
            {
                final_filename = arg.substr(3);
            }
            else if (arg.substr(0, 3) == "-i=")
            {
                if (((arg.length() > 5) && (0 == arg.compare(arg.length() - 5, 5, ".ldpl"))) || ((arg.length() > 4) && (0 == arg.compare(arg.length() - 4, 4, ".lsc"))))
                {
                    if (files_to_compile.size() > 0)
                    {
                        warning("passing multiple LDPL source files to the\ncompiler is deprecated and may be removed in the future.\nPlease use the IMPORT statement instead.");
                    }
                    files_to_compile.push_back(arg.substr(3));
                }
                else
                {
                    // pass everything but .ldpl and .lsc files to the c++ compiler
                    extensions.push_back(arg.substr(3)); // kill -i= prefix
                }
            }
            else if (arg.substr(0, 3) == "-f=")
            {
                // pass flags to the c++ compiler (for example -f=-lSDL)
                extension_flags.push_back(arg.substr(3));
            }
            else if (arg == "-c")
            {
                files_to_compile.push_back(arg);
            }
            else
            {
                cout << "Unknown option: " << arg << endl;
                cout << "Try 'ldpl -h' for more information." << endl;
                return 0;
            }
        }
    }

    // Add default initialization code to the generated source
    state.add_code("int main(int argc, char *argv[]){");
    state.add_code("cout.precision(numeric_limits<ldpl_number>::digits10);");

    // Add default variable declaration code to the generated code
    state.variables[""]["ARGV"] = {2, 3}; // List of text
    state.add_var_code("ldpl_list<chText> " + fix_identifier("ARGV", true) + ";");
    state.variables[""]["ERRORCODE"] = {1}; // Declared in ldpl_lib.cpp
    state.variables[""]["ERRORTEXT"] = {2}; // Declared in ldpl_lib.cpp
    state.add_code("for(int i = 1; i < argc; ++i)");
    state.add_code(fix_identifier("ARGV", true) + ".inner_collection.push_back(argv[i]);");

    // Fail if no filename was passed
    if (files_to_compile.size() == 0)
        error("no source file specified.");

    // For each filename passed, compile each file into one big code
    for (string &filename : files_to_compile)
    {
        // Reset state section for this file
        state.section_state = 0;
        // If the filename was a filename or the user requested standard input
        if (filename != "-c")
        {
            bullet_msg("Loading " + filename);
            load_and_compile(filename, state);
        }
        else
        {
            bullet_msg("Waiting for standard input...");
            accept_and_compile(state);
        }
        // Fail if no procedure section was found
        if (state.section_state < 2)
            error("PROCEDURE section not found" + (filename == "-c" ? "." : " in file '" + filename + "'."));
    }

    // Add return code to the generated main function
    state.add_code("return 0; \n}");

    // Add include for extensions
    if (!extensions.empty())
    {
        for (string &extension : extensions)
        {
            bullet_msg("Including C++ extension " + extension);
            state.add_code("#include \"" + extension + "\"");
        }
    }

    // If an expected subprocedure was not declared, raise an error
    if (state.expected_subprocedures.size() > 0)
        error("the subprocedure " + state.expected_subprocedures[0].first + " is called but never declared.");

    // If only to print the generated code (IR) was required
    if (show_ir)
    {
        cout << "#include \"" << LDPLLIBLOCATION << "/ldpl_lib.cpp\"" << endl; // Add LDPL library
        for (string line : state.variable_code)
            cout << line << endl;
        for (string line : state.subroutine_code)
            cout << line << endl;
        for (string line : state.output_code)
            cout << line << endl;
        exit(0);
    }

    // Otherwise, save the generated code
    ofstream myfile;
    myfile.open("ldpl-temp.cpp");
    myfile << "#include \"" << LDPLLIBLOCATION << "/ldpl_lib.cpp\"" << endl; // Add LDPL library
    for (string line : state.variable_code)
        myfile << line << endl;
    for (string line : state.subroutine_code)
        myfile << line << endl;
    for (string line : state.output_code)
        myfile << line << endl;
    myfile.close();

    // Generate output filename if not set by -o=
    if (final_filename.empty())
    {
        for (unsigned int i = 0; i < output_filename.size(); ++i)
        {
            if (output_filename[i] != '.')
                final_filename += output_filename[i];
            else
                break;
        }
        if (final_filename.size() == 0)
            final_filename = "ldpl-output";
        final_filename += "-bin";
    }

    // Generate the C++ compilation command
    string compile_line = "c++ ldpl-temp.cpp -std=gnu++11 -w -o " + final_filename;
#ifdef STATIC_BUILDS
    if (!no_static)
        compile_line += " -static-libgcc -static-libstdc++ ";
#endif
    if (!extension_flags.empty())
    {
        for (string &flag : extension_flags)
        {
            bullet_msg("using C++ switch " + flag);
            compile_line += " " + flag;
        }
    }

    // Compile the C++ code
    bullet_msg("Building " + final_filename);
    int compiled = system(compile_line.c_str());
#if defined(_WIN32)
    system("del ldpl-temp.cpp");
#else
    system("rm ldpl-temp.cpp");
#endif

    // Output if the code was compiled or not
    if (compiled == 0)
    {
        bullet_msg("Saved as " + final_filename);
        bullet_msg("\033[32;1mFile(s) compiled successfully.\033[0m");
    }
    else
    {
        error("compilation failed.");
    }
}
